contrib/lkt/language/parser.py

```diff
@@ -3024,15 +3024,21 @@ class BinOp(Expr):
 
     @langkit_property()
     def expr_context_free_type():
+        left_cft = Var(Self.left.as_entity.expr_context_free_type)
+        right_cft = Var(Self.right.as_entity.expr_context_free_type)
+
         return Cond(
             # If op is a relational operator, the cf type is bool
             Self.op.is_a(Op.alt_and, Op.alt_or, Op.alt_lt, Op.alt_gt,
                          Op.alt_lte, Op.alt_gte, Op.alt_eq),
 
             Self.bool_type,
 
-            # Else, no idea for the moment. NOTE: we can probably improve that
-            # by forwarding the type of operands.
+            # If the operands types match, the context-free type is the same
+            left_cft.matches(right_cft),
```

Comment:
left_cft._.matches(right_cft)

Reply:
Fixed, still one BinOp missing for `generic_fn_instantiation`, under investigation.

Reply:
This is why this BinOp is missing:

```
[LANGKIT.PROPERTIES] Expr.expr_type_impl (<BinOp test.lkt:78:18-78:27>, <StructDecl prelude: "Int">):
   [LANGKIT.PROPERTIES] BinOp.expr_context_free_type (<BinOp test.lkt:78:18-78:27>):
      [LANGKIT.PROPERTIES] Result: Properties_Error
   [LANGKIT.PROPERTIES] Result: Properties_Error
```

Calling `expr_context_free_type` on the left operand of:

```
test.lkt:78:18: error: Cannot find entity `xxx` in this scope
78 |         (yyy) => xxx * yyy
   |                  ^^^
```

returns a `Properties_Error`, not catched by parser.py while performing type resolution. Therefore, `expr_type_impl` cannot type the `BinOp` since `expr_context_free_type` doesn't return `None` anymore, but propagates the exception, leading to the missing `BinOp` in the output. I'm not sure if we should treat this case as the diagnostic hereabove seems enough (cannot find the type of `xxx` since it's not in the scope of the `BinOp`).

Reply:
As discussed live, this is completely acceptable. Closing the conversation.